<a href="https://colab.research.google.com/github/rick1270/DS-Unit-1-Sprint-4-Statistical-Tests-and-Experiments/blob/master/Copy_of_DS_Unit_1_Sprint_Challenge_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 4

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats
credit_URL = 'https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data'
credit_data = pd.read_csv(credit_URL, header = None, names = ['A1', 'A2', 'A3','A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'A10', 'A11', 'A12', 'A13', 'A14', 'A15', 'A16'])
credit_data.replace('?', np.nan, inplace = True)
credit_data.replace(['f', '-', 'a', 't', '+', 'b'], [0, 0, 0, 1, 1, 1], inplace = True)
credit_data.head()



A1     A2     A3 A4 A5 A6 A7    A8  A9  A10  A11  A12 A13    A14  A15  A16
0  1.0  30.83  0.000  u  g  w  v  1.25   1    1    1    0   g  00202    0    1
1  0.0  58.67  4.460  u  g  q  h  3.04   1    1    6    0   g  00043  560    1
2  0.0  24.50  0.500  u  g  q  h  1.50   1    0    0    0   g  00280  824    1
3  1.0  27.83  1.540  u  g  w  v  3.75   1    1    5    1   g  00100    3    1
4  1.0  20.17  5.625  u  g  w  v  1.71   1    0    0    0   s  00120    0    1

In [3]:
credit_data.describe(include ='all')

A1     A2          A3   A4   A5   A6   A7          A8  \
count   678.000000    678  690.000000  684  684  681  681  690.000000   
unique         NaN    349         NaN    3    3   14    9         NaN   
top            NaN  22.67         NaN    u    g    c    v         NaN   
freq           NaN      9         NaN  519  519  137  399         NaN   
mean      0.690265    NaN    4.758725  NaN  NaN  NaN  NaN    2.223406   
std       0.462725    NaN    4.978163  NaN  NaN  NaN  NaN    3.346513   
min       0.000000    NaN    0.000000  NaN  NaN  NaN  NaN    0.000000   
25%       0.000000    NaN    1.000000  NaN  NaN  NaN  NaN    0.165000   
50%       1.000000    NaN    2.750000  NaN  NaN  NaN  NaN    1.000000   
75%       1.000000    NaN    7.207500  NaN  NaN  NaN  NaN    2.625000   
max       1.000000    NaN   28.000000  NaN  NaN  NaN  NaN   28.500000   

                A9         A10        A11         A12  A13    A14  \
count   690.000000  690.000000  690.00000  690.000000  690    677   
unique         NaN         NaN        NaN         NaN    3    170   
top            NaN         NaN        NaN         NaN    g  00000   
freq           NaN         NaN        NaN         NaN  625    132   
mean      0.523188    0.427536    2.40000    0.457971  NaN    NaN   
std       0.499824    0.495080    4.86294    0.498592  NaN    NaN   
min       0.000000    0.000000    0.00000    0.000000  NaN    NaN   
25%       0.000000    0.000000    0.00000    0.000000  NaN    NaN   
50%       1.000000    0.000000    0.00000    0.000000  NaN    NaN   
75%       1.000000    1.000000    3.00000    1.000000  NaN    NaN   
max       1.000000    1.000000   67.00000    1.000000  NaN    NaN   

                  A15         A16  
count      690.000000  690.000000  
unique            NaN         NaN  
top               NaN         NaN  
freq              NaN         NaN  
mean      1017.385507    0.444928  
std       5210.102598    0.497318  
min          0.000000    0.000000  
25%          0.000000    0.000000  
50%          5.000000    0.000000  
75%        395.500000    1.000000  
max     100000.000000    1.000000

In [4]:
credit_data.isna().sum()

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

In [5]:
credit_data.dtypes

A1     float64
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9       int64
A10      int64
A11      int64
A12      int64
A13     object
A14     object
A15      int64
A16      int64
dtype: object

In [18]:
credit_data[["A2", "A3", "A8", "A11", "A14", "A15", "A1", "A9", "A10", "A12", "A16"]] = credit_data[["A2", "A3", "A8", "A11", "A14", "A15", "A1", "A9", "A10", "A12", "A16"]].astype(float)
credit_data.dtypes

#A1: b, a.
#A2: continuous.
#A3: continuous.
#A4: u, y, l, t.
#A5: g, p, gg.
#A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
#A7: v, h, bb, j, n, z, dd, ff, o.
#A8: continuous.
#A9: t, f.
#A10: t, f.
#A11: continuous.
#A12: t, f.
#A13: g, p, s.
#A14: continuous.
#A15: continuous.
#A16: +,- (class attribute)

A1     float64
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9     float64
A10    float64
A11    float64
A12    float64
A13     object
A14    float64
A15    float64
A16    float64
dtype: object

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" between them and the class label, and apply the Chi-squared test to them. There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end. This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [28]:
class_0 = credit_data[credit_data['A16'] == 0]
class_0.describe()

A1          A2          A3          A8          A9         A10  \
count  374.000000  373.000000  383.000000  383.000000  383.000000  383.000000   
mean     0.700535   29.808231    3.839948    1.257924    0.201044    0.224543   
std      0.458637   10.919291    4.337662    2.120481    0.401305    0.417827   
min      0.000000   15.170000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000   22.000000    0.835000    0.125000    0.000000    0.000000   
50%      1.000000   27.330000    2.210000    0.415000    0.000000    0.000000   
75%      1.000000   34.830000    5.000000    1.500000    0.000000    0.000000   
max      1.000000   80.250000   26.335000   13.875000    1.000000    1.000000   

              A11         A12          A14          A15    A16  
count  383.000000  383.000000   376.000000   383.000000  383.0  
mean     0.631854    0.443864   199.699468   198.605744    0.0  
std      1.900049    0.497489   181.564835   671.608839    0.0  
min      0.000000    0.000000     0.000000     0.000000    0.0  
25%      0.000000    0.000000   100.000000     0.000000    0.0  
50%      0.000000    0.000000   167.500000     1.000000    0.0  
75%      0.000000    1.000000   272.000000    67.000000    0.0  
max     20.000000    1.000000  2000.000000  5552.000000    0.0

In [29]:
class_1 = credit_data[credit_data['A16'] == 1]
class_1.describe()

A1          A2          A3          A8          A9         A10  \
count  304.000000  305.000000  307.000000  307.000000  307.000000  307.000000   
mean     0.677632   33.720492    5.904951    3.427899    0.925081    0.680782   
std      0.468154   12.809641    5.471485    4.120792    0.263690    0.466935   
min      0.000000   13.750000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000   23.170000    1.500000    0.750000    1.000000    0.000000   
50%      1.000000   30.500000    4.460000    2.000000    1.000000    1.000000   
75%      1.000000   41.330000    9.520000    5.000000    1.000000    1.000000   
max      1.000000   76.750000   28.000000   28.500000    1.000000    1.000000   

              A11         A12         A14            A15    A16  
count  307.000000  307.000000  301.000000     307.000000  307.0  
mean     4.605863    0.475570  164.421927    2038.859935    1.0  
std      6.320242    0.500218  161.770675    7659.763941    0.0  
min      0.000000    0.000000    0.000000       0.000000    1.0  
25%      0.000000    0.000000    0.000000       0.000000    1.0  
50%      3.000000    0.000000  120.000000     221.000000    1.0  
75%      7.000000    1.000000  280.000000    1209.000000    1.0  
max     67.000000    1.000000  840.000000  100000.000000    1.0

In [10]:
class_0.mean()

A1       0.700535
A2      29.808231
A3       3.839948
A8       1.257924
A9       0.201044
A10      0.224543
A11      0.631854
A12      0.443864
A14    199.699468
A15    198.605744
A16      0.000000
dtype: float64

In [11]:
class_1.mean()

A1        0.677632
A2       33.720492
A3        5.904951
A8        3.427899
A9        0.925081
A10       0.680782
A11       4.605863
A12       0.475570
A14     164.421927
A15    2038.859935
A16       1.000000
dtype: float64

In [12]:
credit_tests ={}
cat = ['A1', 'A2', 'A3', 'A8', 'A9', 'A10', 'A11', 'A12', 'A14', 'A15', 'A16']
for i in range (1, len(cat)):
  credit_tests[cat[i]] = stats.ttest_ind(class_0[cat[i]], class_1[cat[i]],nan_policy='omit')
  

  
for credit in credit_tests.keys():
  print (credit, credit_tests[credit])

A2 Ttest_indResult(statistic=-4.2922156166315535, pvalue=2.0276637071781407e-05)
A3 Ttest_indResult(statistic=-5.52998337614816, pvalue=4.551680702308068e-08)
A8 Ttest_indResult(statistic=-8.935819983773698, pvalue=3.6710537401601785e-18)
A9 Ttest_indResult(statistic=-27.245398535467622, pvalue=1.9368208799738917e-111)
A10 Ttest_indResult(statistic=-13.52517266303669, pvalue=3.941258302842775e-37)
A11 Ttest_indResult(statistic=-11.667004222431277, pvalue=7.957718568079967e-29)
A12 Ttest_indResult(statistic=-0.8299262276960956, pvalue=0.4068683677059408)
A14 Ttest_indResult(statistic=2.6358251986645476, pvalue=0.008586135473979569)
A15 Ttest_indResult(statistic=-4.680216020964486, pvalue=3.4520256956287944e-06)
A16 Ttest_indResult(statistic=-inf, pvalue=0.0)


In [47]:
A4 = pd.crosstab(credit_data ['A4'], credit_data['A16'])
A4

A16  0.0  1.0
A4           
l      0    2
u    263  256
y    118   45

In [48]:
A6 = pd.crosstab(credit_data ['A6'], credit_data['A16'])
A6

A16  0.0  1.0
A6           
aa    35   19
c     75   62
cc    12   29
d     23    7
e     11   14
ff    46    7
i     45   14
j      7    3
k     37   14
m     22   16
q     27   51
r      1    2
w     31   33
x      6   32

In [49]:
A7 = pd.crosstab(credit_data ['A7'], credit_data['A16'])
A7

A16  0.0  1.0
A7           
bb    34   25
dd     4    2
ff    49    8
h     51   87
j      5    3
n      2    2
o      1    1
v    230  169
z      2    6

In [53]:
from scipy.stats import chisquare
chisquare(A4)

Power_divergenceResult(statistic=array([273.27559055, 365.96039604]), pvalue=array([4.55994382e-60, 3.40964923e-80]))

In [54]:
chisquare(A6)

Power_divergenceResult(statistic=array([192.88888889, 192.08250825]), pvalue=array([3.92856583e-34, 5.74696418e-34]))

In [55]:
chisquare(A7)

Power_divergenceResult(statistic=array([1029.33333333,  792.23762376]), pvalue=array([6.95132317e-217, 9.69179520e-166]))

In [56]:
chisquare(A2)

Power_divergenceResult(statistic=array([11.55479009,  9.72706273]), pvalue=array([0.0090748 , 0.02103454]))

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels? A11 is the only test that rejected the null hypothesis  Ttest_indResult(statistic=-11.667004222431277, pvalue=7.957718568079967e-29)
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?  A6 supports the null hypothosis while A7 appears to reject
- What was the most challenging part of this sprint challenge?  Hardest part was reading chi square

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.